File: tests/trees.py

import enum
from dataclasses import dataclass, replace

class Color(enum.Enum):
    RED   = enum.auto()
    BLACK = enum.auto()


@dataclass
class Node:
    color: Color
    value: int
    left:  'Node' = None
    right: 'Node' = None

    @staticmethod
    def blacken(node) -> 'Node':
        if node is None: return None
        return replace(node, color = Color.BLACK)

    def balance(self) -> 'Node':
        if self.color is Color.RED: return self

        if self.left is not None and \
           self.left.color is Color.RED and \
           self.left.right is not None and \
           self.left.right.color is Color.RED:

           return Node(Color.RED,
                self.left.right.value,
                left = Node(Color.BLACK,
                    self.left.value,
                    left  = self.left.left,
                    right = self.left.right.left,
                ),
                right = Node(Color.BLACK,
                    self.value,
                    left  = self.left.right.right,
                    right = self.right,
                ),
           )
        
        if self.right is not None and \
           self.right.color is Color.RED and \
           self.right.right is not None and \
           self.right.right.color is Color.RED:

            return Node(Color.RED,
                self.right.value,
                left = Node(Color.BLACK,
                    self.value,
                    left  = self.left,
                    right = self.right.left,
                ),
                right = Node.blacken(self.right.right),
            )

        if self.right is not None and \
           self.right.color is Color.RED and \
           self.right.left is not None and \
           self.right.left.color is Color.RED:

           return Node(Color.RED,
                self.right.left.value,
                left = Node(Color.BLACK,
                    self.value,
                    left  = self.left,
                    right = self.right.left.left,
                ),
                right = Node(Color.BLACK,
                    self.right.value,
                    left  = self.right.left.right,
                    right = self.right.right,
                ),
           )
        
        if self.left is not None and \
           self.left.color is Color.RED and \
           self.left.left is not None and \
           self.left.left.color is Color.RED:

            return Node(Color.RED,
                self.left.value,
                left  = Node.blacken(self.left.left),
                right = Node(Color.BLACK,
                    self.value,
                    left  = self.left.right,
                    right = self.right,
                ),
            )

        return self



def insert(node, value: int) -> 'Node':
    if node is None: return Node(Color.RED, value, None, None)

    elif value < node.value:
        return replace(node, left = insert(node.left, value)).balance()
    elif value == node.value:
        return node
    else: # value > node.value:
        return replace(node, right = insert(node.right, value)).balance()


def count_red(node) -> int:
    if node is None:
        return 0
    elif node.color is Color.RED:
        return 1 + count_red(node.left) + count_red(node.right)
    else: # node.color is Color.BLACK
        return count_red(node.left) + count_red(node.right)


if __name__ == "__main__":

    tree = None

    while (value := int(input())) != 0:
        tree = Node.blacken(insert(tree, value))

    print(count_red(tree))

File: tests/trees.exs

defmodule Trees do

	def blacken(node) do
		case node do
			{:Red, data} -> {:Blk, data}
			other        -> other
		end
	end

	def get_value(node) do
		case node do
			{:Red, data} -> data.value
      {:Blk, data} -> data.value
		end
	end

	def map_data(node, f) do
		case node do
			{:Red, data} -> {:Red, f.(data)}
			{:Blk, data} -> {:Blk, f.(data)}
		end
	end

	def balance(tree = {:Blk, data}) do
		case data do
			%{l: {:Red, l = %{r: {:Red, lr}}}} ->
				{:Red, %{
					value: lr.value,
					l: {:Blk, %{
						value: l.value,
						l: l.l,
						r: lr.l, 
					}},
					r: {:Blk, %{
						value: data.value,
						l: lr.r,
						r: data.r,
					}},
				}}

			%{r: {:Red, r = %{r: {:Red, _}}}} ->
				{:Red, %{
					value: r.value,
					l: {:Blk, %{
						value: data.value,
						l: data.l,
						r: r.l,
					}},
					r: blacken(r.r),
				}}

			%{r: {:Red, r = %{l: {:Red, rl}}}} ->
        {:Red, %{
					value: rl.value,
					l: {:Blk, %{
						value: data.value,
						l: data.l,
						r: rl.l,
					}},
					r: {:Blk, %{
						value: r.value,
						l: rl.r, 
						r: r.r,
					}},
				}}

			
			%{l: {:Red, l = %{l: {:Red, _}}}} ->
				{:Red, %{
					value: l.value,
					l: blacken(l.l),
					r: {:Blk, %{
						value: data.value,
						r: data.r,
						l: l.r,
					}},
				}}

      _ -> tree
		end
	end

	def balance(tree), do: tree

	def insert(val, tree) do
		case tree do
			:Empty -> {:Red, %{value: val, l: :Empty, r: :Empty}}
			_ ->
				tree_val = get_value(tree)
				cond do
					val < tree_val ->
						balance(tree |> map_data(fn data -> %{ data | l: insert(val, data.l) } end))

					val == tree_val -> tree

					val > tree_val ->
						balance(tree |> map_data(fn data -> %{ data | r: insert(val, data.r) } end))
				end
		end
	end

  def count_red(:Empty), do: 0
  def count_red({:Red, data}) do
    1 + count_red(data.l) + count_red(data.r)
  end
  def count_red({:Blk, data}) do
    count_red(data.l) + count_red(data.r)
  end


	def get_input() do
		IO.binread(:line)
			|> String.trim()
			|> String.to_integer
	end

	def loop(tree) do
    n = get_input()
    if n == 0 do
      IO.puts(inspect(count_red(tree)))
    else
      loop(blacken(insert(n, tree)))
    end
  end

end

Trees.loop(:Empty)